dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java

```diff
@@ -1323,21 +1356,21 @@ synchronized String evaluateInterval() {
                 currentStats.addDataPoint(throughput, controllerCycle);
             }
 
-            boolean lowTput = false;
-            if (queueDepth == 0 && throughput < poolSize * lowTputThreadsRatio) {
-                lowTput = true;
+            boolean lowActivity = false;
+            if (queueEmpty && ((throughput < (poolSize * lowTputThreadsRatio)) || (activeThreads < (poolSize * activeThreadsGrowthRatio)))) {
```

Comment:
I am not sure if this is the right comparison for lowActivity related to activeThreads.  I would expect it to be something like activeThreads < poolSize - poolIncrement or maybe 2 * poolIncrement.

Reply:
Also need to consider looking if less than coreThreads or currentMinimumPoolSize.  May treat those differently.

Reply:
I am also not sure about this :-) but was just looking for a clear indicator of "low activity". Finding that a sizable chunk (default 25%) of the threads were not doing anything at the moment of the snapshot seemed to qualify.
At this point in the flow, I don't think we need to worry about the absolute number of active threads, e.g. by comparing to coreThreads or currentMinimumPoolSize. This is just to inform getShrinkScore, getGrowScore, and forceVariation that "the pool isn't doing much work right now", so they can make their decisions accordingly. Those methods deal with boundary limits like coreThreads and currentMinimumPoolSize.

Reply:
One of my issues that I have is that 25% means different things at different thread pool sizes.  75% of 12 is 9.  If it is at 8, but the increment level is 8, we go down to 4 when we shouldn't.  We will just to right back up.  This is why I suggested using a factor of the poolDecrement in order to make the decision of being low activity.

When we get to the lower numbers and closer to the coreThreads or currentMinimumPoolSize, we may have to have different characteristics.  Maybe need a Math.max(75% of current pool size, currentMinimumPoolSize);  that is why I brought up having a different criteria when we get close to currentMinimumPoolSize or coreThreads.  I didn't want us to not decide to shrink because we think we have high activity even though we don't because it is not less than 75% of the current pool size, but it is lower than currentMinimumPoolSize / coreThreads.

Reply:
1) The way lowActivity is used in the controller is primarily to curb unnecessary growth - it does not force the pool to shrink. If lowActivity is true, then getGrowScore returns 0 and forceVariation will not grow the pool. The effect of lowActivity on shrink decisions is much more subtle - a nudge to increase shrinkScore in some cases. 

2)  We don't have to worry about lowActivity pushing the pool size below currentMinimumPoolSize because that is prevented by checks in adjustPoolSize.

3) The example you gave will never arise, because poolDecrement will always be no greater than Math.floor(25% of pool size).

For example, say cpus=2 - then coreThreads is 4 and poolIncrement/Decrement is 2 
So if poolSize=12, then activeThreads < 9 (8 or lower) would help set lowActivity, but dropping 2 threads from 12 leaves 10, which is 2 above activeThreads.
Things are closer on the first increment - if poolSize=6 then activeThreads has to be 4 or lower for lowActivity to be set, and a poolDecrement of 2 is still safe

4) To overcome any concern about shrinking below activeThreads, it is probably a good idea to add it as a limit to shrinking in adjustPoolSize:

            if (poolAdjustment < 0 && newPoolSize >= Math.max(currentMinimumPoolSize,activeThreads) {

That would protect us against unexpected behavior if the various system properties have been manually configured in such a way that poolDecrement might be greater than 25% of poolSize. 
~~~~~~~~
The thing that I like about 75% is that it provides a very general flag that "hey the pool is really not very busy at the moment" which is a good clue that the controller really should not grow the pool, even randomly, and should think about maybe shrinking it. :-)